xds/src/main/java/io/grpc/xds/LookasideLb.java

```diff
@@ -49,169 +48,132 @@
+  @Nullable
+  private XdsConfig xdsConfig;
+  // Most recent EndpointWatcher.
+  // Becomes non-null once handleResolvedAddresses() successfully.
+  @Nullable
+  private EndpointWatcher endpointWatcher;
+
+  @Nullable
+  private ObjectPool<XdsClient> cachedXdsClientRef; // Cached for optimization.
+  @Nullable
+  private LoadReportClient cachedLrsClient; // Cached for optimization.
+
+  LookasideLb(Helper lookasideLbHelper, EndpointUpdateCallback endpointUpdateCallback) {
     this(
         lookasideLbHelper,
-        edsUpdateCallback,
+        endpointUpdateCallback,
         LoadBalancerRegistry.getDefaultRegistry(),
         LocalityStoreFactory.getInstance(),
-        LoadReportClientFactory.getInstance());
+        LoadReportClientFactory.getInstance(),
+        Bootstrapper.getInstance());
   }
 
   @VisibleForTesting
   LookasideLb(
       Helper lookasideLbHelper,
-      EdsUpdateCallback edsUpdateCallback,
+      EndpointUpdateCallback endpointUpdateCallback,
       LoadBalancerRegistry lbRegistry,
       LocalityStoreFactory localityStoreFactory,
-      LoadReportClientFactory loadReportClientFactory) {
-    this.edsUpdateCallback = edsUpdateCallback;
+      LoadReportClientFactory loadReportClientFactory,
+      Bootstrapper bootstrapper) {
+    this.lookasideLbHelper = lookasideLbHelper;
+    this.channelLogger = lookasideLbHelper.getChannelLogger();
+    this.endpointUpdateCallback = endpointUpdateCallback;
     this.lbRegistry = lbRegistry;
-    this.lookasideChannelLb = new GracefulSwitchLoadBalancer(lookasideLbHelper);
+    this.switchingLoadBalancer = new GracefulSwitchLoadBalancer(lookasideLbHelper);
     this.localityStoreFactory = localityStoreFactory;
     this.loadReportClientFactory = loadReportClientFactory;
-  }
-
-  @Override
-  protected LoadBalancer delegate() {
-    return lookasideChannelLb;
+    this.bootstrapper = bootstrapper;
   }
 
   @Override
   public void handleResolvedAddresses(ResolvedAddresses resolvedAddresses) {
+    channelLogger.log(ChannelLogLevel.DEBUG, "Received ResolvedAddresses '%s'", resolvedAddresses);
+
     // In the future, xdsConfig can be gotten directly by
-    // resolvedAddresses.getLoadBalancingPolicyConfig()
+    // resolvedAddresses.getLoadBalancingPolicyConfig().
     Attributes attributes = resolvedAddresses.getAttributes();
-    Map<String, ?> newRawLbConfig = checkNotNull(
-        attributes.get(ATTR_LOAD_BALANCING_CONFIG), "ATTR_LOAD_BALANCING_CONFIG not available");
+    Map<String, ?> newRawLbConfig = attributes.get(ATTR_LOAD_BALANCING_CONFIG);
+    if (newRawLbConfig == null) {
+      // This will not happen when the service config error handling is implemented.
+      // For now simply go to TRANSIENT_FAILURE.
+      lookasideLbHelper.updateBalancingState(
+          TRANSIENT_FAILURE,
+          new ErrorPicker(
+              Status.UNAVAILABLE.withDescription("ATTR_LOAD_BALANCING_CONFIG not available")));
```

Comment:
Do you need to return here? Otherwise giving `null` `newRawLbConfig` to `parseLoadBalancingConfigPolicy` could generate an NPE, right?

Reply:
Yes, thanks for the catch.